### PR TITLE
Jakarta Enterprise Beans 4.0 Final

### PR DESCRIPTION
**Pull request by jeanouii**
Signed-off-by: Jean-Louis Monteiro <jlmonteiro@tomitribe.com>

## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [x] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [x] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [x] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [x] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [x] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable.
- [x] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - [x] Generated IP Log
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [x] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/repositories/jakartaejb-1003/jakarta/ejb/jakarta.ejb-api/4.0.0/
- [x] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      [Covered by Platform TCK](https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-jakartaeetck-9.0.0-RC1.zip)
- [x] The URL of the compatibility certification request issue:
      [Enterprise Beans Certification Request]( https://github.com/eclipse-ee4j/jakartaee-platform/issues/269)
- [x] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

### CODE REVIEW

**Comment by starksm64** [comment]
# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request
 
10. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by starksm64** [comment]
@jeanouii let me known when the initial PR template items are complete.

**Comment by jeanouii** [comment]
Just generated and updated the apidocs and spec documents in case someone wants to start reviewing.

**Comment by dblevins** [comment]
Note to mentors that we do need to get https://github.com/eclipse-ee4j/jakartaee-tck/pull/452 merged before we can proceed to a final ballot.  Not anticipating any issues, just FYI.

**Comment by starksm64** [comment]
Note that I have added checklist items for the EMO/PMC tasks. The https://projects.eclipse.org/projects/ee4j.ejb/releases/4.0 project page needs the date updated to something in the future so that the review can be scheduled.

**Comment by jeanouii** [comment]
@starksm64 added the link to API and Javadoc staging repo
Also fixed the release date in the page you mentioned.

File: enterprise-beans/4.0/apidocs/META-INF/MANIFEST.MF
Comment (hussainnm) [inline enterprise-beans/4.0/apidocs/META-INF/MANIFEST.MF:1]: Please remove the META-INF folder from apidocs

**Comment by hussainnm** [comment]
@jeanouii The date in specification document is in french locale, can you re-generate the documents by setting maven locale to en_US.
```
MAVEN_OPTS="-Duser.language=en -Duser.region=US"
mvn --version
mvn -Pstaging clean package "-Dstatus=Final" "-Ddocversion=4.0"
```

**Comment by starksm64** [comment]
@jeanouii I have updated the checklist to verify the current contents. At this point we just need the platform PR to be able to use the TCK and CCR to complete this.

**Comment by jeanouii** [comment]
Thanks @starksm64

**Comment by scottmarlow** [comment]
FYI, the RC1 Platform TCK link is https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-jakartaeetck-9.0.0-RC1.zip which has the same SHA `75b2493a117e7f8fe775cc7e2e8a605a203611895091f4bb9aaabed57f813392` as was used [to pass 100% of the Full Platform TCK yesterday](https://ci.eclipse.org/jakartaee-tck/job/jakartaee-tck/job/master/1165) (we are running again due to a system networking failure that occurred after the test run while collecting test results).  The [Jakarta EE 9 TCK results wiki page](https://github.com/eclipse-ee4j/jakartaee-tck/wiki/Jakarta-EE-9-TCK-results) will be updated after the run completes, not that you need to wait since I said, we already passed 100%.

[We also passed the Web Profile tests as well!](https://ci.eclipse.org/jakartaee-tck/job/eftl-jakartaeetck-certification-web/19/junit-reports-with-handlebars/testSuitesOverview.html).

This testing was done with https://download.eclipse.org/ee4j/glassfish/glassfish-6.0.0-RC2.zip.

**Review by ivargrimstad — CHANGES_REQUESTED** [review]
Some things to fix before ballot:
- Link to compatible implementation is missing.
- Link to certification request is missing
- Link to TCK and signature
- Link to artifact in maven central

**Review by ivargrimstad — CHANGES_REQUESTED** [review]
Some thing to fix before starting ballot:
- Add link to compatible implementation (in _index.md)
- Add link to Certification request (in PR)
- Add TCK links (zip, sig, sha, pub) (in _index.md)
- Add link to artifact in Maven Central (in _index.md)

**Comment by hussainnm** [comment]
The final jar needs to be re-staged as it is referring to jakarta.transaction-api-2.0.0-RC1

https://jakarta.oss.sonatype.org/content/repositories/jakartaejb-1002/jakarta/ejb/jakarta.ejb-api/4.0.0/jakarta.ejb-api-4.0.0.pom

**Comment by jeanouii** [comment]
@hussainnm New build done and staging link updated.
@kwsutter @ivargrimstad Did the changes, if you can double check quickly

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [inline enterprise-beans/4.0/_index.md:16]: These links should be the final resting location of the TCK:  https://download.eclipse.org/jakartaee/platform/9/jakarta-jakartaeetck-9.0.0.zip (just like you have for the sig and sha files).

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [inline enterprise-beans/4.0/_index.md:61]: Cesar's name shouldn't be listed since David and Jean-Louis have voting rights.  This can be corrected when the ballot is updated.

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [inline enterprise-beans/4.0/_index.md:80]: Cesar's name shouldn't be listed since David and Jean-Louis have voting rights.  This can be corrected when the ballot is updated.

File: enterprise-beans/4.0/apidocs/META-INF/MANIFEST.MF
Comment (kwsutter) [inline enterprise-beans/4.0/apidocs/META-INF/MANIFEST.MF:1]: The apidocs directory should not have a META-INF directory (part of the checklist).  Please remove.

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [inline enterprise-beans/4.0/_index.md:51]: You need to list a Compatible Implementation.  In the case of EJB, you should be referencing the RC2 of Glassfish (same as your Compatibility Request).

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [inline enterprise-beans/4.0/_index.md:9]: You need to complete and schedule a release review per the checklist in the opening Comment on this PR.  I also pinged you about this on Slack. Thanks!

**Review by kwsutter — COMMENTED** [review]
Nothing major, but some of these items need to be cleaned up for the ballot.  I'm not ready to Approve it yet...

File: enterprise-beans/4.0/apidocs/index.html
Comment (scottmarlow) [inline enterprise-beans/4.0/apidocs/index.html:10]: Should the enterprise-beans/4.0/apidocs/index.html contain javascript?  Or would html be better?

Comment (kwsutter) [reply enterprise-beans/4.0/apidocs/index.html:10]: I guess it doesn't affect the displaying of the javadoc, does it?  At least I couldn't find a visible issue with the javadoc...

**Comment by kwsutter** [comment]
I just verified most of the items in the secondary checklist (https://github.com/jakartaee/specifications/pull/258#issuecomment-681928971).  If the other items in my review can be cleaned up, then I think we're ready for a ballot.  Thanks!

File: enterprise-beans/4.0/_index.md
Comment (jeanouii) [reply enterprise-beans/4.0/_index.md:16]: Done

File: enterprise-beans/4.0/_index.md
Comment (jeanouii) [reply enterprise-beans/4.0/_index.md:61]: Done

File: enterprise-beans/4.0/_index.md
Comment (jeanouii) [reply enterprise-beans/4.0/_index.md:80]: Done

File: enterprise-beans/4.0/apidocs/META-INF/MANIFEST.MF
Comment (jeanouii) [reply enterprise-beans/4.0/apidocs/META-INF/MANIFEST.MF:1]: Done

File: enterprise-beans/4.0/_index.md
Comment (jeanouii) [reply enterprise-beans/4.0/_index.md:51]: Done

File: enterprise-beans/4.0/apidocs/index.html
Comment (jeanouii) [reply enterprise-beans/4.0/apidocs/index.html:10]: I picked up the result of the api javadoc jar.
But I can try to generate again and see?

Comment (dblevins) [reply enterprise-beans/4.0/apidocs/index.html:10]: We don't have a "no javascript" rule, so I'm not sure as to why this is relevant.

**Comment by edbratt** [comment]
Prior to this going to ballot, could you please remove the Milestone ballot results from _index.md? That briefly caused me some confusion since it suggests this is already approved.

**Comment by dblevins** [comment]
@edbratt I didn't see a milestone ballot.  There was a plan review ballot, which @jeanouii removed, but I think we actually need to keep the plan review.  We did for Activation for example: https://jakarta.ee/specifications/activation/2.0/

File: enterprise-beans/4.0/_index.md
Comment (dblevins) [reply enterprise-beans/4.0/_index.md:9]: EMO/PMC tasks should be done now.  I don't know if I did them correctly, but I expect the EMO will provide guidance if not.

**Comment by dblevins** [comment]
Ok I see most specs did remove the plan review from the page.  I think we need to bring this up on the Specification Committee as we do need to keep that data publicly visible somewhere.

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [reply enterprise-beans/4.0/_index.md:9]: Thanks.

**Comment by kwsutter** [comment]
> Ok I see most specs did remove the plan review from the page. I think we need to bring this up on the Specification Committee as we do need to keep that data publicly visible somewhere. We don't need to resolve that immediately. There's likely some work we need to do.

I would say that most Specs did not remove that material from the _index.md file.  There were some comments in the _index.md file in case you did hold your own review, but most Specs just had to point at the main plan review.  So, some Specs did remove that Comment section I had put in place.

For EJB, you did have your own Plan Review and that ballot material needs to stay in the _index.md file.  We can correct this when we do the ballot update (if we don't restore that material before we go to ballot).

**Comment by kwsutter** [comment]
> Prior to this going to ballot, could you please remove the Milestone ballot results from _index.md? That briefly caused me some confusion since it suggests this is already approved.

That material is valid for EJB since they held their own Plan Review.  We should not have removed it.  If it gets restored before we go to ballot, great.  If not, it needs to be restored when we update the _index.md for the final ballot.

File: enterprise-beans/4.0/_index.md
Comment (kwsutter) [inline enterprise-beans/4.0/_index.md:45]: This was not a proper update.  EJB held it's own Plan Review and we had a ballot that needs to be counted for.  This won't hold up our ballot, but needs to be restored post-ballot.

**Review by kwsutter — APPROVED** [review]
One minor item that we can clean up after the ballot.  I'm going to leave this open and let @ivargrimstad  take a final look.  If all is good, he can start the ballot during his morning hours.

File: enterprise-beans/4.0/apidocs/index.html
Comment (scottmarlow) [reply enterprise-beans/4.0/apidocs/index.html:10]: It displays correctly for me also (using Firefox).  I mentioned it in case it slipped by without notice as unlikely as that was.

Comment (scottmarlow) [reply enterprise-beans/4.0/apidocs/index.html:10]: > I picked up the result of the api javadoc jar.
> But I can try to generate again and see?

I tried generating locally and see the same javascript (which displayed correctly in firefox).

**Comment by starksm64** [comment]
I have gone through the checklist and generated site and it looks good. I agree with @kwsutter that the plan review vote should be left in.

**Comment by edbratt** [comment]
I'm happy with the way _index.md looks now. If the previous review ballot details need to be kept, can we do anything to clairify so it doesn't appear that the ballot is complete. Perhaps add the empty final ballot table and/or put them in reverse order. I hadn't seen that in other Specs so that caught me off guard (it appeared as if the ballot had already been run). I agree most Specs. didn't have separate plan reviews. Sorry if this caused needless confusion.

**Comment by edbratt** [comment]
I don't think there is any need to include a GlassFish Platform Cert. request link here. This Spec. doesn't require the Platform certification.

File: enterprise-beans/4.0/_index.md
Comment (ivargrimstad) [reply enterprise-beans/4.0/_index.md:45]: Agree!. Let's restore the plan review section when we post the ballot results.

**Review by ivargrimstad — APPROVED** [review]

**Comment by ivargrimstad** [comment]
> I don't think there is any need to include a GlassFish Platform Cert. request link here. This Spec. doesn't require the Platform certification.

I'll remove it from the checklist and initiate the ballot.

**Comment by starksm64** [comment]
* On ballot completion, the specification committee mentor:
 - [x] adds this final checklist to the main PR.
 - [] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [x] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job.
    N/A - platform project
 - [] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [] creates an issue in the specification project that includes the following checklist for the specification project team:
 N/A

**Comment by starksm64** [comment]
Representative | Representative for: | Vote
-- | -- | --
Kenji Kazumura | Fujitsu | +1
Dan Bandera, Kevin Sutter | IBM | +1
Ed Bratt, Dmitry Kornilov | Oracle | +1
Andrew Pielage, Matt Gill | Payara | +1
Scott Stark, Mark Little | Red Hat | +1
David Blevins, Jean-Louis Monteiro | Tomitribe | +1
Ivar Grimstad | EE4J PMC | +1
Marcelo Ancelmo, Martijn Verburg | Participant Members | +1
Werner Keil | Committer Members | +1
Scott (Congquan) Wang | Enterprise Members | +1
  | Total | 10
  |   |  
  |   |  
Non-binding votes | Role | Vote
Hussain N M |   | +1

**Pull request merged**
